onnxruntime/core/platform/windows/device_discovery.cc

```diff
@@ -3,46 +3,275 @@
+        if (vendor_id == 0 && device_id == 0) {
+          continue;
+        }
+
+        key = GetDeviceKey(vendor_id, device_id);
+
+        if (device_info.find(key) == device_info.end()) {
+          device_info[key] = {};
+        } else {
+          if (guid == GUID_DEVCLASS_PROCESSOR) {
+            // skip duplicate processor entries as we don't need to accumulate bus numbers for them
+            continue;
+          }
+        }
+
+        entry = &device_info[key];
+        entry->vendor_id = vendor_id;
+        entry->device_id = device_id;
+      } else {
+        // need valid ids
+        continue;
+      }
+
+      // Get device description.
+      if (SetupDiGetDeviceRegistryPropertyW(devInfo, &devData, SPDRP_DEVICEDESC, nullptr,
+                                            (PBYTE)buffer.data(), (DWORD)buffer.size(), &size)) {
+        entry->description = buffer;
+
+        // Should we require the NPU to be found by DXCore or do we want to allow this vague matching?
+        // Probably depends on whether we always attempt to run DXCore or not.
+        const auto possible_npu = [](const std::wstring& desc) {
+          return (desc.find(L"NPU") != std::wstring::npos ||
+                  desc.find(L"Neural") != std::wstring::npos ||
+                  desc.find(L"AI Engine") != std::wstring::npos ||
+                  desc.find(L"VPU") != std::wstring::npos);
+        };
+
+        // not 100% accurate. is there a better way?
+        uint64_t npu_key = GetDeviceKey(*entry);
+        bool is_npu = npus.count(npu_key) > 0 || possible_npu(entry->description);
+
+        if (guid == GUID_DEVCLASS_DISPLAY) {
+          entry->type = OrtHardwareDeviceType_GPU;
+        } else if (guid == GUID_DEVCLASS_PROCESSOR) {
+          entry->type = is_npu ? OrtHardwareDeviceType_NPU : OrtHardwareDeviceType_CPU;
+        } else if (guid == GUID_DEVCLASS_SYSTEM) {
+          if (!is_npu) {
+            // we're only iterating system devices to look for NPUs so drop anything else
+            device_info.erase(key);
+            continue;
+          }
+
+          entry->type = OrtHardwareDeviceType_NPU;
+        } else {
+          // unknown device type
+          continue;
+        }
+      } else {
+        // can't set device type
+        continue;
+      }
+
+      if (SetupDiGetDeviceRegistryPropertyW(devInfo, &devData, SPDRP_MFG, nullptr,
+                                            (PBYTE)buffer.data(), (DWORD)buffer.size(), &size)) {
+        entry->vendor = buffer;
+      }
+
+      if (guid != GUID_DEVCLASS_PROCESSOR) {
+        DWORD busNumber = 0;
+        size = 0;
+        if (SetupDiGetDeviceRegistryPropertyW(devInfo, &devData, SPDRP_BUSNUMBER, nullptr,
+                                              reinterpret_cast<PBYTE>(&busNumber), sizeof(busNumber), &size)) {
+          // push_back in case there are two identical devices. not sure how else to tell them apart
```

Comment:
I could be completely wrong, but would [https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nf-setupapi-setupdigetdeviceinstanceidw](SetupDiGetDeviceInstanceIdW) be useful here?

>The following is an example of an instance ID ("1&08") concatenated to a device ID for a PCI device:
>
>PCI\VEN_1000&DEV_0001&SUBSYS_00000000&REV_02\1&08

Reply:
Consider all this code to be a temporary/interim solution so there are probably lots of things that could be better.

I did start with this, but couldn't find a way to join a unique identifier from the SetupApi world with the DX LUID. Best I could find was vendor_id + device_id (which seems to be good for PCI but hit-and-miss for ACPI), so the info I'm getting from SetupApi is hardware type specific not instance specific (hence the list of BusNumber if you have say 2 identical GPUs in your system). 

and once vendo id + device id was the info joining SetupApi with DX, calling SetupDiGetDeviceRegistryPropertyW didn't seem to be adding any value (although it's quite possible I have overlooked something and it can add). 

so the comment is a bit misleading and should be updated.